scripts/generate_upgrade_syncset.py

```diff
@@ -0,0 +1,301 @@
+
+        data['spec'] = {
+            'clusterDeploymentRefs': [
+                {
+                    'name': dq(cluster_name),
+                }
+            ],
+            'resourceApplyMode': 'Upsert',
+            'resources': [
+                uc
+            ],
+        }
+        items.append(data)
+
+    outer['items'] = items
+
+    yaml.preserve_quotes = True
+    yaml.dump(outer, sys.stdout, Dumper=yaml.RoundTripDumper)
+
+
+def init_argparse():
+    """
+    Initialises the command-line argument parser
+    :return: parser instance
+    """
+    parser = argparse.ArgumentParser(description='UpgradeConfig SelectorSyncSet generator',
+                                     usage='%(prog)s [options]')
+
+    parser.add_argument('--subgroup', required=False, choices=['all', 'prod', 'nonprod'], default='all',
+                        help='Sub-cluster-group to target (default=all)')
+    parser.add_argument('--start-time', required=False, type=valid_date,
+                        help='Timestamp at which upgrade should commence (eg. 2020-01-25T00:00:00Z)')
+    parser.add_argument('--version', required=True, type=valid_version, help='Target upgrade version [required]')
+    parser.add_argument('--channel', required=True, help='Target upgrade channel [required]')
+    parser.add_argument('--force', '-f', required=False, default=False, help='Force upgrade [optional, default false]')
+    parser.add_argument('--environment', '-e', required=False, choices=['production', 'staging', 'integration'],
+                        default='production', help='Hive environment [optional, default production]')
+
+    cluster_group = parser.add_argument_group()
+    cluster_group.add_argument('--cluster-id', '-c', required=False, action='append', help='Cluster ID to target')
+    cluster_group.add_argument('--group', '-g', required=False, action='append',
+                               choices=['sd', 'sre', 'cicd', 'internal', 'snowflake', 'external'],
+                               help='Cluster group to target')
+
+    return parser
+
+
+def valid_version(v):
+    """
+    Verify the supplied version adheres to formatting conventions
+    :param v: version to verify
+    :return: the unmodified version if it parses successfully
+    :raises ArgumentError if version is invalid
+    """
+    p = re.compile('^4\\.[0-9]+\\.[0-9]+')
+    result = p.match(v)
+    if not result:
+        raise argparse.ArgumentError('Version must be in format 4.X.Y, eg. 4.3.25')
+    return v
+
+
+def valid_date(d):
+    """
+    Verify the supplied timestamp adheres to formatting conventions
+    :param d: timestamp to verify
+    :return: the unmodified timestamp if it parses successfully
+    :raises ArgumentError if timestamp is invalid
+    """
+    try:
+        datetime.datetime.strptime(d, '%Y-%m-%dT%H:%M:%SZ')
+    except ValueError:
+        raise argparse.ArgumentError('Time must be in format YYYY-MM-DDTHH:MM:SSZ, eg. 2020-01-25T00:00:00Z')
+    return d
```

Comment:
Why is this needed?

Reply:
The date is going straight into the UpgradeConfig CR as the scheduled upgrade time in UTC, and needs to be in ISO-8601 format when it hits the CR. So this is just forcing the user of the script to put the correct format in initially to avoid the SyncSet pushing something out with a bad timestamp.